### ISSUE TITLE
SEGFAULT in the process of stopping the context

### ISSUE THREAD

**Issue by Arenoros** [warmcat/libwebsockets#2974]
When using libuv without foreign_loop, after calling lws_context_destroy from the context thread, the already deleted handle may be added to the endgame_handles queue due to accessing already freed memory

I think the cause is this uv_run call in `lib/event-libs/libuv/libuv.c:elops_destroy_context1_uv`

``` cpp
while (budget-- && (m = uv_run(pt_to_priv_uv(pt)->io_loop,
                UV_RUN_NOWAIT)))
        ;
```
Because all descriptors that were added to the cleanup queue when `lib\event-libs\libuv\libuv.c:lws_io_cb` was called, including the one for which `lws_io_cb` was called, will be deleted...

![image](https://github.com/warmcat/libwebsockets/assets/2578918/5aca80a2-c282-4451-b23f-6c6848d42768)

![image](https://github.com/warmcat/libwebsockets/assets/2578918/88e8e3b8-01c3-49e2-bd1e-f604a7a4c312)

...in this uv_run call:
![image](https://github.com/warmcat/libwebsockets/assets/2578918/ab92a111-2f77-46a8-b073-76c5463b2fc5)

Because of this, when returning to `uv__fast_poll_process_poll_poll_req`, the handle that has already been cleared will be checked and depending on our luck we will either crash or terminate successfully.
![image](https://github.com/warmcat/libwebsockets/assets/2578918/6e143eb5-c3ad-4fc5-b6c8-f8777becdd85)

**Comment by Arenoros** [comment]
I have so far just commented out these 3 lines and it solved my crashing problem but maybe I misunderstood the actual cause

**Comment by lws-team** [comment]
Getting the event loop library support working well for all the options was a bit of a trek... after spending several days testing and cleaning the various examples after the last work on it I came to realize the built-in loop support was a bad idea for loops like libuv that have deferred close and I should deprecate it in future, although I left it working for the examples AFAIK.

Foreign loops are OK because finalizing everything so you can delete the loop is done outside of lws and after lws itself has logically closed everything, the foreign loop may also not want to close just because lws is detaching and that's OK too. One-shot closure event libs are OK with internal loops too because we can clean up fully before trying to close the loop.  But this awkward mix where lws is also responsible to close the event loop as well as itself, and event loop handles close asynchronously, is a step too far with this ugly workaround to crank the event loop manually to try to finish up the closures before closing the loop.  It works for the examples AFAIK but it's not hard to imagine the event lib changing something internally that would make it blow up.

So I'm afraid there's no clean solution except switch to the foreign loop model, the main code will be unchanged just you need to instantiate lws context after creating it and destroy the lws context before destroying it.

**Comment by Arenoros** [comment]
Thanks for the answer, just in case I want to clarify, I understand correctly that for correct termination of lws_service regardless of using eventlib as foreign loop or using built-in event loop in lws should be called lws_context_destroy from the event loop thread, аfter that, calling lws_context_destroy again will already destroy the context object?

**Comment by lws-team** [comment]
Whatever is in the examples should match what the lws code thinks you should do, generally for different lws versions (how this this internal and external loop destroy works differs quite a lot by lws version).

In recent lws, the context destroy is stateful and split into stages which in the simplest case (no event lib) just happen sequentially in one call.  The idea behind two calls is the loop will continue naturally after the first has closed everything from lws and set a 'we're shutting down' flag, when all handled on the loop finish their async close and are destroyed, the loop will exit and the second destroy call actually destroy things (including the loop object in internal case).

**Comment by Arenoros** [comment]
I'll formulate the question a bit more precisely, if I'm using the current version 4.3.2 and using the lws event loop is it safe to call lws_context_destroy in the event loop or in this case it should be called only after exiting lws_service? While I was looking for the reason of the crash I looked through all available examples and quite a bit of the lws_context_destroy call itself, but I still didn't get the full confidence in this question.

**Comment by lws-team** [comment]
The reason for your specific crash goes through whatever version of libuv you're using and the recursion situation for that.

As I said the flow is simplified by using foreign loops, I suggest you switch to that and see if you still have a problem.

**Comment by Arenoros** [comment]
There seems to be no problem with using foreign loops, I just wanted to clarify some unclear points with stopping and removing context.
